hydra-node/src/Hydra/Chain/Direct/State.hs

```diff
@@ -0,0 +1,400 @@
+    } ->
+    HydraStateMachine 'StClosed
+
+deriving instance Show (HydraStateMachine st)
+
+getKnownUTxO ::
+  OnChainHeadState st ->
+  UTxO
+getKnownUTxO OnChainHeadState{stateMachine} =
+  case stateMachine of
+    Idle{} ->
+      mempty
+    Initialized{initialThreadOutput, initialInitials, initialCommits} ->
+      UTxO $
+        Map.fromList $
+          take2Of4 initialThreadOutput : (take2Of3 <$> (initialInitials <> initialCommits))
+    Open{openThreadOutput = (i, o, _, _)} ->
+      UTxO.singleton (i, o)
+    Closed{closedThreadOutput = (i, o, _, _)} ->
+      UTxO.singleton (i, o)
+
+-- Working with opaque states
+
+-- | An existential wrapping /some/ 'OnChainHeadState' into a value that carry
+-- no type-level information about the state.
+data SomeOnChainHeadState where
+  SomeOnChainHeadState ::
+    forall st.
+    HasTransition st =>
+    OnChainHeadState st ->
+    SomeOnChainHeadState
+
+-- | Some Kind for witnessing Hydra state-machine's states at the type-level.
+--
+-- This is useful to
+--
+-- (a) Reads code evolving the state machine, as it makes transition more
+-- obvious from type-signatures;
+-- (b) Pattern-match on the 'HydraStateMachine' without having to bother with
+-- non-reachable cases.
+data HeadStateKind = StIdle | StInitialized | StOpen | StClosed
+
+-- | A token witnessing the state's type of an 'OnChainHeadState'. See 'reifyState'
+data TokHeadState (st :: HeadStateKind) where
+  TkIdle :: TokHeadState 'StIdle
+  TkInitialized :: TokHeadState 'StInitialized
+  TkOpen :: TokHeadState 'StOpen
+  TkClosed :: TokHeadState 'StClosed
+
+deriving instance Show (TokHeadState st)
+
+-- | Reify a 'HeadStateKind' kind into a value to enable pattern-matching on
+-- existentials.
+reifyState :: forall st. OnChainHeadState st -> TokHeadState st
+reifyState OnChainHeadState{stateMachine} =
+  case stateMachine of
+    Idle{} -> TkIdle
+    Initialized{} -> TkInitialized
+    Open{} -> TkOpen
+    Closed{} -> TkClosed
+
+-- Initialization
+
+-- | Initialize a new 'OnChainHeadState'.
+idleOnChainHeadState ::
+  NetworkId ->
+  VerificationKey PaymentKey ->
+  Party ->
+  OnChainHeadState 'StIdle
+idleOnChainHeadState networkId ownVerificationKey ownParty =
+  OnChainHeadState
+    { networkId
+    , ownVerificationKey
+    , ownParty
+    , stateMachine = Idle
+    }
+
+-- Constructing Transitions
+
+-- NOTE / TODO: The only real reason why all the function belows are in STM is
+-- because they do access the wallet's UTXO for error reporting. This is
+-- misleading and slightly annoying for testing given that all those functions
+-- are actually "pure" (modulo exceptions which can be captured as 'Either').
+--
+-- I'd suggest to remove the `MonadThrow` instance and instead have the
+-- potentially failing functions return an 'Either' and, do the exception
+-- throwing in the upper-layer, wrapping the error with details such as the
+-- wallet UTxO.
```

Comment:
This NOTE/TODO seems out of date?

Reply:
Ah yes. I did that already :sweat_smile: !